### PR TITLE
Added async support for com.mongodb.MongoClientSettings

### PR DESCRIPTION
**Pull request by rozza**
Deprecated async.client.MongoClientSettings
Unifies configuration for the async and sync packages

JAVA-2837

### CODE REVIEW

File: driver-async/src/main/com/mongodb/async/client/MongoClientSettings.java [hunk 1]
```diff
@@ -365,13 +356,16 @@ public Builder readConcern(final ReadConcern readConcern) {
          * @param credentialList the credential list
          * @return this
          * @see MongoClientSettings#getCredentialList()
-         * @deprecated Prefer {@link #credential(MongoCredential)}
+         * @deprecated support for multiple credentials has been removed use {@link #credential(MongoCredential)} instead
```
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/MongoClientSettings.java]: Change here: we no longer support multiple credentials

File: driver-sync/src/main/com/mongodb/client/MongoClientImpl.java [hunk 2]
```diff
@@ -121,17 +122,18 @@ private static Cluster createCluster(final MongoClientSettings settings,
...
-                settings.getConnectionPoolSettings(), getStreamFactory(settings), getStreamFactory(settings), credentialList,
+                settings.getConnectionPoolSettings(), getStreamFactory(settings, false), getStreamFactory(settings, true), credentialList,
                 getCommandListener(settings.getCommandListeners()), settings.getApplicationName(), mongoDriverInformation,
                 settings.getCompressorList());
     }
 
-    private static StreamFactory getStreamFactory(final MongoClientSettings settings) {
+    private static StreamFactory getStreamFactory(final MongoClientSettings settings, final boolean isHeartbeat) {
         StreamFactoryFactory streamFactoryFactory = settings.getStreamFactoryFactory();
+        SocketSettings socketSettings = isHeartbeat ? settings.getHeartbeatSocketSettings() : settings.getSocketSettings();
```
Comment (rozza) [inline driver-sync/src/main/com/mongodb/client/MongoClientImpl.java]: Fixes bug - previously same socket settings used for all scenrios.

File: driver-async/src/main/com/mongodb/async/client/MongoClients.java [hunk 3]
```diff
@@ -175,13 +217,14 @@ public static CodecRegistry getDefaultCodecRegistry() {
         return com.mongodb.MongoClientSettings.getDefaultCodecRegistry();
     }
 
-    private static StreamFactory getStreamFactory(@Nullable final StreamFactoryFactory streamFactoryFactory,
-                                                  final SocketSettings socketSettings, final SslSettings sslSettings,
-                                                  final String streamType) {
+    private static StreamFactory getStreamFactory(final com.mongodb.MongoClientSettings settings, final String streamType,
+                                                  final boolean isHeartbeat) {
+        StreamFactoryFactory streamFactoryFactory = settings.getStreamFactoryFactory();
+        SocketSettings socketSettings = isHeartbeat ? settings.getHeartbeatSocketSettings() : settings.getSocketSettings();
```
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/MongoClients.java]: Fixes bug - previously same socket settings used for all scenrios.

**Comment by rozza** [comment]
@jyemin updated as per conversation.

File: docs/reference/content/driver-async/getting-started/quick-start.md [hunk 4]
```diff
@@ -134,7 +134,7 @@ The connection string mostly follows [RFC 3986](http://tools.ietf.org/html/rfc39
 MongoClient mongoClient = MongoClients.create(new ConnectionString("mongodb://localhost"));
 ```
 
-- You can call [`MongoClients.create()`]({{< apiref "com/mongodb/async/client/MongoClients.html#create-com.mongodb.async.client.MongoClientSettings-" >}}) with a [`MongoClientSettings`]({{< apiref "com/mongodb/async/client/MongoClientSettings.html">}}) object:
+- You can call [`MongoClients.create()`]({{< apiref "com/mongodb/async/client/MongoClients.html#create-com.mongodb.async.client.MongoClientSettings-" >}}) with a [`MongoClientSettings`]({{< apiref "com/mongodb/MongoClientSettings.html">}}) object:
```
Comment (jyemin) [inline docs/reference/content/driver-async/getting-started/quick-start.md]: The first apiref (for the create method) also needs to be updated.

File: driver-async/src/main/com/mongodb/async/client/MongoClientSettings.java [hunk 5]
```diff
@@ -74,46 +76,36 @@ public static Builder builder(final MongoClientSettings settings) {
...
-        private final com.mongodb.MongoClientSettings.Builder wrappedBuilder = com.mongodb.MongoClientSettings.builder();
         private SocketSettings.Builder heartbeatSocketSettingsBuilder = null;
 
         private Builder() {
+            wrappedBuilder = com.mongodb.MongoClientSettings.builder();
+        }
+
+        private Builder(final com.mongodb.MongoClientSettings settings) {
+            notNull("settings", settings);
+            wrappedBuilder = com.mongodb.MongoClientSettings.builder(settings);
```
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/MongoClientSettings.java]: Looks like a credential in the wrapped MongoClientSettings will be ignored.  I think this.credentialList has to be initialized with the wrappedBuilder.credential if it's non-null.

File: driver-async/src/main/com/mongodb/async/client/MongoClients.java [hunk 6]
```diff
@@ -121,35 +129,72 @@ public static MongoClient create(final MongoClientSettings settings, @Nullable f
...
-                mongoDriverInformation, settings.getCompressorList()), externalResourceCloser);
+        return new MongoClientImpl(settings, createCluster(settings, mongoDriverInformation, streamFactory, heartbeatStreamFactory),
+                externalResourceCloser);
+    }
+
+    @SuppressWarnings("deprecation")
+    private static Cluster createCluster(final MongoClientSettings settings, @Nullable final MongoDriverInformation mongoDriverInformation,
+                                         final StreamFactory streamFactory, final StreamFactory heartbeatStreamFactory) {
+        notNull("settings", settings);
+        List<MongoCredential> credentialList = settings.getCredential() != null ? Collections.singletonList(settings.getCredential())
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-async/src/main/com/mongodb/async/client/MongoClients.java]: Good catch.

**Pull request closed without merging**